cmd/note_common.go

```diff
@@ -145,17 +143,83 @@ func createCommitComments(project string, mrID int, commit string, body string,
 	// tracking information (new line & old line number pairs,
 	// and file information)
 	scanner := bufio.NewScanner(strings.NewReader(body))
+	lastDiffLine := ""
+	comments := ""
 	newfile := ""
 	oldfile := ""
-	oldLineNum := -1
-	newLineNum := -1
-	comments := ""
+
 	for scanner.Scan() {
-		if strings.HasPrefix(scanner.Text(), "| newfile:") {
-			if comments != "" {
-				createCommitNote(project, mrID, commit, newfile, oldfile, oldLineNum, newLineNum, comments, block)
-				comments = ""
+		line := scanner.Text()
+
+		if !strings.HasPrefix(line, "| ") {
+			comments += "\n" + line
+			continue
+		}
+
+		if comments != "" && lastDiffLine == "" {
+			log.Fatal("Cannot comment on first line of commit (context unknown).")
+		}
+
+		if comments != "" {
+			linetype := ""
+			oldLineNum := -1
+			newLineNum := -1
+
+			// parse lastDiffLine
+			f := strings.Fields(strings.TrimSpace(lastDiffLine))
+
+			// The output can be, for example:
+			// | # # [no +/-] < context comment
+			// | # #    +     < newline comment
+			// | # #    -     < oldline comment
+			// | #      -     < oldline comment
+			// |   #    +     < newline comment
+
+			// f[0] is always a "|"
+			// f[1] will always be a number
+			val1, _ := strconv.Atoi(f[1])
+			val2, err := strconv.Atoi(f[2])
+
+			if err == nil { // f[2] is a number
+				if len(f) <= 3 { // f[3] does not exist
+					oldLineNum = val1
+					newLineNum = val2
+					linetype = "context"
+				} else {
+					newLineNum = val2
+					switch {
+					case strings.HasPrefix(f[3], "+"):
+						linetype = "new"
+					case strings.HasPrefix(f[3], "-"):
+						linetype = "old"
+					default:
+						linetype = "context"
+					}
+				}
+			} else { // f[2] is not a number
+				switch {
+				case strings.HasPrefix(f[2], "+"):
+					newLineNum = val1
+					linetype = "new"
+				case strings.HasPrefix(f[2], "-"):
+					oldLineNum = val1
+					linetype = "old"
+				default:
+					panic("unknown string in diff")
+				}
 			}
+
+			createCommitNote(project, mrID, commit, newfile, oldfile, linetype, oldLineNum, newLineNum, comments, block)
+			comments = ""
+		}
+
+		f := strings.Fields(strings.TrimSpace(line))
+		if f[1] == "@@" {
+			lastDiffLine = ""
+			continue
+		}
+
+		if f[1] == "newfile:" {
 			// read filename
 			f := strings.Split(scanner.Text(), " ")
```

Comment:
Just minor nit: since we already have `line` at this point we wouldn't need to call for `scanner.Text()`.